### PR TITLE
Wip heapster

### PR DESCRIPTION
**Pull request by brianhuffman**
This PR is not intended to be merged; rather it gives us a way to view the changes between the `wip-heapster` and `master` branches, and to discuss which of those changes should be made into saw-core PRs.

### CODE REVIEW

File: saw-core/prelude/Prelude.sawcore [hunk 1]
```diff
@@ -117,6 +119,14 @@ data Eq (t : sort 1) (x : t) : t -> sort 1 where {
...
   }
 
+-- The version of Eq in Prop
+-- FIXME: this should replace the predicative Eq when we are sure everything
+-- works ok with it
+data EqP (t : sort 1) (x : t) : t -> Prop where {
+    ReflP : EqP t x x;
+  }
+
+
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:129]: @eddywestbrook This change of `Eq` from `sort 0` to `Prop` seems like something that would be reasonable to do in saw-core master. We should open a separate github issue requesting this change.

File: saw-core/prelude/Prelude.sawcore [hunk 2]
```diff
@@ -1389,6 +1474,287 @@ primitive mkDouble : Integer -> Integer -> Float;
...
+-- Axiom: x <=u y <u z implies x <u z
+axiom trans_bvult_bvule : (n:Nat) -> (x y z:bitvector n) ->
+                          is_bvult n x y -> is_bvule n y z -> is_bvult n x z;
+
+axiom bvult_sub_add_bvult : (n:Nat) -> (x y z:bitvector n) ->
+                            is_bvule n y z -> is_bvult n x (bvSub n z y) ->
+                            is_bvult n (bvAdd n y x) z;
+
+-- The type of a bitvector-indexed array of length len : bitvector n
+primitive BVVec : (n:Nat) -> bitvector n -> sort 0 -> sort 0;
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:1553]: I'm puzzled about why this type needs to exist as a primitive. Couldn't you just define it as:
```
BVVec n w a = Vec (bvToNat n w) a
```

File: saw-core/prelude/Prelude.sawcore [hunk 3]
```diff
@@ -1389,6 +1474,287 @@ primitive mkDouble : Integer -> Integer -> Float;
...
+              (bvultWithProof n (bvSub n i start') len'))
+           (atBVVec n len a v i pf));
+
+
+-------------------------------------------------------------------------------
+-- Dependent sums
+
+data Sigma (a : sort 0) (b : a -> sort 0) : sort 0 where {
+    exists : (pa : a) -> b pa -> Sigma a b;
+}
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:1673]: A Sigma type seems like a reasonable addition to the saw-core prelude on the master branch. We should create a github ticket asking for this.

File: saw-core/prelude/Prelude.sawcore [hunk 4]
```diff
@@ -1389,6 +1474,287 @@ primitive mkDouble : Integer -> Integer -> Float;
...
+             (\ (pa:a) (pb: b pa) -> pb);
+
+
+--------------------------------------------------------------------------------
+-- Lists
+
+data List (a:sort 0) : sort 0 where {
+  Nil : List a;
+  Cons : a -> List a -> List a;
+}
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:1696]: A `List` datatype seems like it would make sense in the `master` saw-core prelude. We should make a ticket asking for it.

File: saw-core/prelude/Prelude.sawcore [hunk 5]
```diff
@@ -1401,53 +1767,210 @@ composeM : (a b c: sort 0) -> (a -> CompM b) -> (b -> CompM c) -> a -> CompM c;
 composeM a b c f g x = bindM b c (f x) g;
 
 -- Raise an error in the computation monad
-primitive errorM : (a:sort 0) -> CompM a;
+primitive errorM : (a:sort 0) -> String -> CompM a;
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:1770]: This and other `CompM`-related changes might make sense as a separate PR, given that we already have `CompM` in the saw-core prelude on the master branch.

File: saw-core/prelude/Prelude.sawcore [hunk 6]
```diff
@@ -759,6 +803,32 @@ eqNatPrec : (x y : Nat) -> eqNat (Succ x) (Succ y) -> eqNat x y;
...
+primitive proveEqNat : (m n : Nat) -> Maybe (EqP Nat m n);
+
+-- | Try to prove x <= y (FIXME: implement this from natCompareLe!)
+primitive proveLeNat : (x y : Nat) -> Maybe (IsLeNat x y);
+
+-- | Try to prove x < y
+proveLtNat : (x y : Nat) -> Maybe (IsLtNat x y);
+proveLtNat x y = proveLeNat (Succ x) y;
+
+
```
Comment (brianhuffman) [inline saw-core/prelude/Prelude.sawcore:831]: These propositional `Nat` comparisons would probably make sense in master; we could make a PR for this.

File: saw-core/saw-core.el [hunk 7]
```diff
@@ -0,0 +1,137 @@
+;;; saw-core.el --- A major mode for editing saw-core  -*- lexical-binding: t; -*-
```
Comment (brianhuffman) [inline saw-core/saw-core.el:1]: We should probably make a separate PR that adds `saw-core.el` to the master branch.

File: saw-core/src/Verifier/SAW/Term/Pretty.hs [hunk 8]
```diff
@@ -20,6 +20,7 @@ module Verifier.SAW.Term.Pretty
   , depthPPOpts
   , ppNat
   , ppTerm
+  , ppTermInCtx
```
Comment (brianhuffman) [inline saw-core/src/Verifier/SAW/Term/Pretty.hs:23]: Is `ppTermInCtx` something that would make sense in the master branch? Maybe this should get its own PR.

File: saw-core/prelude/Prelude.sawcore (hunk 1 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:129]: Yes, this makes sense. I didn't want to change the existing definition of `Eq` as it seemed like it could be a breaking change for something and I wasn't sure how to check that it didn't break anything, but I do think it should be moved into `Prop`, just like it is in Coq.

File: saw-core/prelude/Prelude.sawcore (hunk 2 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:1553]: This certainly makes sense. My one concern is whether there are any definitional equalities that no longer work when defining `BVVec` this way. I think probably not, but I will have to try it out. Maybe I'll add a Heapster ticket to investigate this change, and if it works I'll let you know.

File: saw-core/prelude/Prelude.sawcore (hunk 3 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:1673]: Yes, I agree, that makes sense.

File: saw-core/prelude/Prelude.sawcore (hunk 4 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:1696]: Agreed!

File: saw-core/prelude/Prelude.sawcore (hunk 5 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:1770]: Sure that makes sense. What actions do you need from me?

On a related note: we are going to be making more changes to the `CompM` monad in the next month or so as part of our shift to using itrees to implement `CompM`. Not only will there be new combinators in the prelude, but there will also be a new dependency on the itrees library in `saw-core-coq`. Should we wait on merging these `CompM` changes until the itrees-related changes are resolved? How should we discuss the plan for those changes?

File: saw-core/prelude/Prelude.sawcore (hunk 6 above)
Comment (eddywestbrook) [reply saw-core/prelude/Prelude.sawcore:831]: Agreed

File: saw-core/saw-core.el (hunk 7 above)
Comment (eddywestbrook) [reply saw-core/saw-core.el:1]: Yeah, I think more people might use it. If nothing else, it's better than fundamental mode in emacs...

File: saw-core/src/Verifier/SAW/Term/Pretty.hs (hunk 8 above)
Comment (eddywestbrook) [reply saw-core/src/Verifier/SAW/Term/Pretty.hs:23]: Sure, that makes sense. It just exposes a little more power in the pretty-printing API, nothing that I imagine would break any existing code.

File: saw-core/prelude/Prelude.sawcore (hunk 1 above)
Comment (brianhuffman) [reply saw-core/prelude/Prelude.sawcore:129]: See #154.

File: saw-core/prelude/Prelude.sawcore (hunk 3 above)
Comment (brianhuffman) [reply saw-core/prelude/Prelude.sawcore:1676]: See #155.

File: saw-core/saw-core.el (hunk 7 above)
Comment (brianhuffman) [reply saw-core/saw-core.el:1]: See #157.

File: saw-core/src/Verifier/SAW/Term/Pretty.hs (hunk 8 above)
Comment (brianhuffman) [reply saw-core/src/Verifier/SAW/Term/Pretty.hs:26]: See #167.
